### ISSUE TITLE
Thoughts around enabling modules

### ISSUE THREAD

**Issue by 123andy** [vanderbilt/redcap-external-modules#13]
Currently when you go to enable a module in your server, the external modules presents a modal frame...  I'm thinking that over time there are going to be a LOT of modules and this interface is going to be less than ideal.

What about making it a normal page with a datatable.  This would make for easy searching, etc. Also, we could use the 'expand row' feature ([Data Table Rows](https://datatables.net/examples/api/row_details.html)) to make it easy to view more details on how the module works, etc...

Thoughts?

**Comment by taylorr4** [comment]
@123andy it sounds like a good idea. Definitely will be useful inside projects where more modules will be displayed in the modal.

**Comment by tbembersimeao** [comment]
I think it would be nice to have additional hooks like:
- `hook__module_enable`: at the moment the module is enabled (useful when some module depends on db changes to work, e.g. creating tables)
- `hook_module_disable` at the moment that the module is disabled (useful for garbage cleaning, e.g. removing tables)

**Comment by mmcev106** [comment]
I think adding module event hooks would be very useful.  At the moment I'm envisioning the following:
- hook_module_system_enable
- hook_module_system_disable
- hook_module_system_change_version
- hook_module_project_enable
- hook_module_project_disable

**Comment by tbembersimeao** [comment]
That sounds great!

**Comment by 123andy** [comment]
The idea was floated of being able to add a module-configurable 'validation' page.  Say, for example, a module requested that a new table be added to the database.  When the module is enabled, it could check if that table exists - and if it didn't, it could present the user with the sql for creating the table.

I also wonder about version confirmations - say someone wants to enable a new module - if the module had only been tested at version 7.6.2 and the current version was 7.6.6 the hook might be able to ask the user if they still want to proceed given that the module hadn't been tested against that version...

**Comment by tbembersimeao** [comment]
@mmcev106 Created a PR to add the new hooks: https://github.com/vanderbilt/redcap-external-modules/pull/74.

**Comment by pbchase** [comment]
We could really benefit from this feature. Today we have been designing yet another module that would benefit from it. In each case, the goal is to create a side-table in MySQL to serve the data-storage needs of the module. We want to make modules that are accessible to the non-technical REDCap community, but to do this we need more services from the External Modules framework. 

There are certainly other ways to address the database creation need that are not as far reaching as hooking the module enabling/disabling. For example we could define a database schema in the config.json and have the EM code run the create/alter table statements on behalf of the module. Yet that idea opens up another can of worms. It matters little to us as long as we can meet the need within the EM framework.

How would you like to proceed?

**Comment by mmcev106** [comment]
I fully intend on reviewing and merging the linked pull request.  It will likely be a few more weeks before we can realistically look into this though.  It's is a big change to review and consider possible long term support consequences.  We only have a handful of hours a month available for requests that don't directly relate projects we're working on...

**Comment by pbchase** [comment]
Thank you, @mmcev106. Our need is not urgent, but we would very much like to see something like this happen.

**Comment by mmcev106** [comment]
Closing, since the PR was merged.  It might make sense to continue some of the other discussion in separate issues.
